libsolidity/analysis/TypeChecker.cpp

```diff
@@ -338,45 +340,75 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
 			m_errorReporter.typeError(5587_error, _function.location(), "Internal functions cannot be payable.");
 	}
+
+	vector<VariableDeclaration const*> internalParametersInConstructor;
+
 	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& var) {
 		if (type(var)->category() == Type::Category::Mapping)
 		{
 			if (var.referenceLocation() != VariableDeclaration::Location::Storage)
 			{
-				if (!_function.libraryFunction() && _function.isPublic())
+				if (!_function.libraryFunction() && _function.isPublic() && !_function.isConstructor())
 					m_errorReporter.typeError(3442_error, var.location(), "Mapping types can only have a data location of \"storage\" and thus only be parameters or return variables for internal or library functions.");
 				else
 					m_errorReporter.typeError(5380_error, var.location(), "Mapping types can only have a data location of \"storage\"." );
 			}
 			else
-				solAssert(_function.libraryFunction() || !_function.isPublic(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
+			{
+				if (_function.isConstructor())
+					internalParametersInConstructor.emplace_back(&var);
+				else
+					solAssert(_function.libraryFunction() || !_function.isPublic(), "Mapping types for parameters or return variables can only be used in internal or library functions.");
+			}
 		}
 		else
 		{
-			if (!type(var)->canLiveOutsideStorage() && _function.isPublic())
-				m_errorReporter.typeError(3312_error, var.location(), "Type is required to live outside storage.");
-			if (_function.isPublic())
+			if (!type(var)->canLiveOutsideStorage())
+			{
+				if (_function.isConstructor())
+					internalParametersInConstructor.emplace_back(&var);
+				else if (_function.isPublic())
+					m_errorReporter.typeError(3312_error, var.location(), "Type is required to live outside storage.");
+			}
+			if (
+				(!_function.isConstructor() && _function.isPublic()) ||
+				(_function.isConstructor() && !m_currentContract->abstract())
```

Comment:
Would it make sense to refactor this condition into a function in `FunctionDefinition`?

Reply:
Turned it into a local variable.